### PR TITLE
add support for reading WebDataset inputs

### PR DESCRIPTION
**Pull request by danielmend**
- will allow us to run video2dataset + clip-video-encode end to end

### CODE REVIEW

File: clip_video_encode/clip_video_encode.py [hunk 1]
```diff
@@ -91,26 +119,36 @@ def clip_video_encode(
...
+        vids, ids, meta = reader.get_data()
+        meta_refs = list(range(len(vids)))
+
+    else: # WebDataset, so we distribute shards
+        vids = list(braceexpand.braceexpand(src))
 
     starting_shard_id = 0
-    shard_sample_count = 10000
+    shard_sample_count = 100
+    num_vids_per_input_shard = 1500
```
Comment (iejMac) [inline clip_video_encode/clip_video_encode.py:138]: what is this?

File: clip_video_encode/clip_video_encode.py [hunk 2]
```diff
@@ -91,26 +119,36 @@ def clip_video_encode(
...
-    vids, ids, meta = reader.get_data()
-    meta_refs = list(range(len(vids)))
+
+    if input_format == "table":
+        reader = Reader(src, metadata_columns)
+        vids, ids, meta = reader.get_data()
+        meta_refs = list(range(len(vids)))
+
+    else: # WebDataset, so we distribute shards
+        vids = list(braceexpand.braceexpand(src))
```
Comment (iejMac) [inline clip_video_encode/clip_video_encode.py:134]: aaah maybe let's not call this vids, it's shards, lets not confuse the two

File: clip_video_encode/clip_video_encode.py [hunk 3]
```diff
@@ -41,17 +60,47 @@ def encode_chunk(frames, ind_dict, writer, mapper, preprocess, meta, ids, use_ds
...
 
+def read_shard(tempdir):
+    """
+    Extract video filepaths, video ids, and metadata from the contents of an opened WebDataset shard
+
+    Input:
+        tempdir:
+            path to directory containing contents of an opened WebDataset shard with input data
+    """
+    vids = sorted([f for f in os.listdir(tempdir) if f.endswith(".mp4")])  # TODO: parameterize the video extension
```
Comment (iejMac) [inline clip_video_encode/clip_video_encode.py:80]: glob.glob("*.mp4")

File: clip_video_encode/clip_video_encode.py [hunk 4]
```diff
@@ -120,36 +183,152 @@ def clip_video_encode(
...
+                device,
+            )
+    else:  # WebDataset shard logic
+        shard_times = []
+        for shard in shards:
+            times = {}
+            t = time.time()
+            try:
+                tempdir = tempfile.mkdtemp()
+                os.chmod(tempdir, 0o777)
```
Comment (iejMac) [inline clip_video_encode/clip_video_encode.py:251]: do you really need this? I've never had to do this? Also why .mkdtemp()? what's wrong with the normal ```with tempfile.TemporaryDirectory() as tmpdir:```?

File: clip_video_encode/clip_video_encode.py [hunk 5]
```diff
@@ -120,36 +183,152 @@ def clip_video_encode(
...
+                subprocess.run(["aws", "s3", "cp", shard, tempdir], check=True)
```
Comment (iejMac) [inline clip_video_encode/clip_video_encode.py:252]: tar files don't need to be on s3

Comment (iejMac) [reply clip_video_encode/clip_video_encode.py:252]: but I guess it's fine for now

File: clip_video_encode/clip_video_encode.py [hunk 6]
```diff
@@ -120,36 +183,152 @@ def clip_video_encode(
...
+                        ids,
+                        use_dst_name,
+                        device,
+                    )
+                times["encode"] = times.get("encode", 0) + time.time() - t
+                t = time.time()
+            finally:
+                shutil.rmtree(tempdir)
+                # writer.close()
+                print(f"Frames: {n_frames}")
```
Comment (iejMac) [inline clip_video_encode/clip_video_encode.py:321]: Maybe clean this up

Comment (iejMac) [reply clip_video_encode/clip_video_encode.py:321]: or just make it more simple - one line with the most important performance information

File: clip_video_encode/clip_video_encode.py (hunk 5 above)
Comment (danielmend) [reply clip_video_encode/clip_video_encode.py:235]: should i keep this then? or change it (to fsspec?)

**Pull request merged**